File: android/src/main/java/com/unact/yandexmapkit/CircleMapObjectController.java
package com.unact.yandexmapkit;

import androidx.annotation.NonNull;

import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.CircleMapObject;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectTapListener;

import java.lang.ref.WeakReference;
import java.util.Map;

public class CircleMapObjectController extends MapObjectController implements MapObjectTapListener {
  private final boolean internallyControlled;
  public final CircleMapObject circle;
  private boolean consumeTapEvents = false;
  @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  private final WeakReference<YandexMapController> controller;
  public final String id;

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public CircleMapObjectController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    CircleMapObject circle = parent.addCircle(
      Utils.circleFromJson((Map<String, Object>) params.get("circle"))
    );

    this.circle = circle;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.internallyControlled = false;

    circle.setUserData(this.id);
    circle.addTapListener(this);
    update(params);
  }

  public CircleMapObjectController(
    CircleMapObject circle,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    this.circle = circle;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.internallyControlled = true;

    circle.setUserData(id);
    circle.addTapListener(this);
    update(params);
  }

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public void update(Map<String, Object> params) {
    if (!internallyControlled) {
      circle.setGeometry(Utils.circleFromJson((Map<String, Object>) params.get("circle")));
      circle.setVisible((Boolean) params.get("isVisible"));
    }

    circle.setGeodesic((Boolean) params.get("isGeodesic"));
    circle.setZIndex(((Double) params.get("zIndex")).floatValue());
    circle.setStrokeColor(((Number) params.get("strokeColor")).intValue());
    circle.setStrokeWidth(((Double) params.get("strokeWidth")).floatValue());
    circle.setFillColor(((Number) params.get("fillColor")).intValue());

    consumeTapEvents = (Boolean) params.get("consumeTapEvents");
  }

  public void remove() {
    if (internallyControlled) {
      return;
    }

    circle.getParent().remove(circle);
  }

  @Override
  public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectTap(id, point);

    return consumeTapEvents;
  }
}

File: ios/Classes/CircleMapObjectController.swift
import YandexMapsMobile

class CircleMapObjectController: NSObject, MapObjectController, YMKMapObjectTapListener {
  private let internallyControlled: Bool
  public let circle: YMKCircleMapObject
  private var consumeTapEvents: Bool = false
  public weak var controller: YandexMapController?
  public let id: String

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let circle = parent.addCircle(with: Utils.circleFromJson(params["circle"] as! [String: Any]))

    self.circle = circle
    self.id = params["id"] as! String
    self.controller = controller
    self.internallyControlled = false

    super.init()

    circle.userData = self.id
    circle.addTapListener(with: self)
    update(params)
  }

  public required init(
    circle: YMKCircleMapObject,
    params: [String: Any],
    controller: YandexMapController
  ) {
    self.circle = circle
    self.id = params["id"] as! String
    self.controller = controller
    self.internallyControlled = true

    super.init()

    circle.userData = self.id
    circle.addTapListener(with: self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    if (!internallyControlled) {
      circle.geometry = Utils.circleFromJson(params["circle"] as! [String: Any])
      circle.isVisible = (params["isVisible"] as! NSNumber).boolValue
    }

    circle.isGeodesic = (params["isGeodesic"] as! NSNumber).boolValue
    circle.zIndex = (params["zIndex"] as! NSNumber).floatValue
    circle.strokeColor = Utils.uiColor(fromInt: (params["strokeColor"] as! NSNumber).int64Value)
    circle.strokeWidth = (params["strokeWidth"] as! NSNumber).floatValue
    circle.fillColor = Utils.uiColor(fromInt: (params["fillColor"] as! NSNumber).int64Value)

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    if (internallyControlled) {
      return
    }

    circle.parent.remove(with: circle)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller!.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }
}
